In [1]:
import numpy as np
import pandas as pd
import pingouin as pg

## Soru 1

In [2]:
name = ['Gonul', 'Eda', 'Mustafa', 'Kemal', 'Ayse', 'Metin'] 
# gender --> nominal
gender = ['Female','Female','Male','Male','Female','Male']
# age --> ratio
age = [27, 24, 22, 32, 35, 29]
# apartment_no --> interval
apartment_no = ['5th', '1th', '2th', '4th', '3th', '6th']
# floor --> ordinal
floor = ['Top Floor', 'Ground Floor', 'Ground Floor', 'Second Floor', 'Second Floor', 'Top Floor']
Occupants = {"Name ": name, "Gender ": gender, "Age ": age, "ApartmentNo ": apartment_no, "Floor ": floor}
data = pd.DataFrame(Occupants)
data

Name  Gender   Age  ApartmentNo         Floor 
0    Gonul  Female    27          5th     Top Floor
1      Eda  Female    24          1th  Ground Floor
2  Mustafa    Male    22          2th  Ground Floor
3    Kemal    Male    32          4th  Second Floor
4     Ayse  Female    35          3th  Second Floor
5    Metin    Male    29          6th     Top Floor

**Level of Measurements** 
1. Nominal: Gender kolonu nominal veriler içerir. Kategori içerir ama sıralanamaz.
2. Ordinal: Floor kolonu ordinal veriler içerir. Sıralanabilir ama ground ve top verileri sayısal bir anlam ifade etmez.
3. Interval: ApartmentNo kolonu interval veriler içerir. Natural zero starting point yoktur. Apartmanda ilerledikçe hangi kapı numarasının sonra geldiğini biliriz.
4. Ratio: Age kolonu ratio veriler içerir. True zero point vardır. Belli bir ölçeğe sahiptir.

## Soru 2

In [3]:
# import data file
df = pd.read_csv('data1.csv')
df

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0      842302         M        17.99         10.38          122.80     1001.0   
1      842517         M        20.57         17.77          132.90     1326.0   
2    84300903         M        19.69         21.25          130.00     1203.0   
3    84348301         M        11.42         20.38           77.58      386.1   
4    84358402         M        20.29         14.34          135.10     1297.0   
..        ...       ...          ...           ...             ...        ...   
564    926424         M        21.56         22.39          142.00     1479.0   
565    926682         M        20.13         28.25          131.20     1261.0   
566    926954         M        16.60         28.08          108.30      858.1   
567    927241         M        20.60         29.33          140.10     1265.0   
568     92751         B         7.76         24.54           47.92      181.0   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0            0.11840           0.27760         0.30010              0.14710   
1            0.08474           0.07864         0.08690              0.07017   
2            0.10960           0.15990         0.19740              0.12790   
3            0.14250           0.28390         0.24140              0.10520   
4            0.10030           0.13280         0.19800              0.10430   
..               ...               ...             ...                  ...   
564          0.11100           0.11590         0.24390              0.13890   
565          0.09780           0.10340         0.14400              0.09791   
566          0.08455           0.10230         0.09251              0.05302   
567          0.11780           0.27700         0.35140              0.15200   
568          0.05263           0.04362         0.00000              0.00000   

     ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0    ...          17.33           184.60      2019.0           0.16220   
1    ...          23.41           158.80      1956.0           0.12380   
2    ...          25.53           152.50      1709.0           0.14440   
3    ...          26.50            98.87       567.7           0.20980   
4    ...          16.67           152.20      1575.0           0.13740   
..   ...            ...              ...         ...               ...   
564  ...          26.40           166.10      2027.0           0.14100   
565  ...          38.25           155.00      1731.0           0.11660   
566  ...          34.12           126.70      1124.0           0.11390   
567  ...          39.42           184.60      1821.0           0.16500   
568  ...          30.37            59.16       268.6           0.08996   

     compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0              0.66560           0.7119                0.2654          0.4601   
1              0.18660           0.2416                0.1860          0.2750   
2              0.42450           0.4504                0.2430          0.3613   
3              0.86630           0.6869                0.2575          0.6638   
4              0.20500           0.4000                0.1625          0.2364   
..                 ...              ...                   ...             ...   
564            0.21130           0.4107                0.2216          0.2060   
565            0.19220           0.3215                0.1628          0.2572   
566            0.30940           0.3403                0.1418          0.2218   
567            0.86810           0.9387                0.2650          0.4087   
568            0.06444           0.0000                0.0000          0.2871   

     fractal_dimension_worst  Unnamed: 32  
0                    0.11890          NaN  
1                    0.08902          NaN  
2                    0.08758          NaN  
3                    0.17300          NaN  
4      

In [4]:
# drop some columns
df = df.drop(['Unnamed: 32','id'], axis = 1)
# view some basic statistical details like mean, std etc.
df.describe()

radius_mean  texture_mean  perimeter_mean    area_mean  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       symmetry_mean  fractal_dimension_mean  ...  radius_worst  \
count     569.000000              569.000000  ...    569.000000   
mean        0.181162                0.062798  ...     16.269190   
std         0.027414                0.007060  ...      4.833242   
min         0.106000                0.049960  ...      7.930000   
25%         0.161900                0.057700  ...     13.010000   
50%         0.179200                0.061540  ...     14.970000   
75%         0.195700                0.066120  ...     18.790000   
max         0.304000                0.097440  ...     36.040000   

       texture_worst  perimeter_worst   area_worst  smoothness_worst  \
count     569.000000       569.000000   569.000000        569.000000   
mean       25.677223       107.261213   880.583128          0.132369   
std         6.146258        33.602542   569.356993          0.022832   
min        12.020000        50.410000   185.200000          0.071170   
25%        21.080000        84.110000   515.300000          0.116600   
50%        25.410000        97.660000   686.500000          0.131300   
75%        29.720000       125.400000  1084.000000          0.146000   
max        49.540000       251.200000  4254.000000          0.222600   

       compactness_worst  concavity_worst  concave points_worst  \
count         569.000000       569.000000            569.000000   
mean            0.254265         0.272188              0.114606   
std             0.157336         0.208624              0.065732   
min             0.027290         0.000000              0.000000   
25%             0.147200         0.114500              0.064930   
50%             0.211900         0.226700              0.099930   
75%             0.339100         0.382900              0.161400   
max             1.058000         1.252000              0.291000   

       symmetry_worst  fractal_dimension_worst  
count      569.000000               569.000000  
mean         0.290076                 0.083946  
std          0.061867                 0.018061  
min          0.156500                 0.055040  
25%          0.250400                 0.071460  
50%          0.282200                 0.080040  
75%          0.317900                 0.092080  
max          0.663800                 0.207500  

[8 rows x 30 columns]

1. Tablonun geneline baktığımız zaman, ilk bakışta Mean ve Median değerleri arasında çok büyük bir fark görünmüyor. Outlier değerler ortalamayı çok saptırmamış. Normal dağılıma yakınsak bir data var elimizde.
2. Std değerini de diğer değerlerle karşılaştırdığımız zaman çok büyük bir farklılık görünmüyor ilk bakışta. Sadece area_mean ve area_worst kolonlarında bütün değerler için yüksek değerler dikkat çekiyor.
3. Q1, Q3 ve Median değerlerini karşılaştırdığımız zaman dağılımların dengeli olduğu gözlenebilir ve bu da verimizin normal dağılımlı bir veri olduğunu göstermektedir.
4. Yine area_mean ve area_worst kolonları için Min/Q1 ve Max/Q3 değerleri arasında yüksek bir farklılık göze çarpıyor.

### **Outliers of Radius Mean**

In [5]:
# separate data in similar situation 
df_bening = df[df["diagnosis"] == "B"]
df_malignant = df[df["diagnosis"] == "M"]

In [6]:
# view the values of variable radius_mean and detect outliers of radius_mean
desc = df_bening.radius_mean.describe()
Q1 = desc[4]
Q3 = desc[6]
IQR = Q3 - Q1
lower_bound = Q1 - 1.5*IQR
upper_bound = Q3 + 1.5*IQR
print("Anything outside this range is an outlier: (", lower_bound, "," , upper_bound, ")")
df_bening[df_bening.radius_mean < lower_bound].radius_mean
print("Outliers: ", df_bening[(df_bening.radius_mean < lower_bound) | (df_bening.radius_mean > upper_bound)].radius_mean.values)

Anything outside this range is an outlier: ( 7.645000000000001 , 16.805 )
Outliers:  [ 6.981 16.84  17.85 ]


* Mean ve median değerleri arasında çok büyük farklılıklar gözlenmemektedir. Normal dağılıma yakınsak verilere sahip olduğumuz söylenebilir.
* lower_bound olan 7.645000000000001 değerinin altında 1 outlierımız, upper_bound olan 16.805 değerimizin üzerinde 2 outlier değerimiz bulunmaktadır.
* Outlier değerlerimizi incelediğimiz zaman çok uç değerlere sahip olmadığını görüyoruz.

### **Outliers of Area Mean**

In [7]:
# view the values of variable area_mean and detect outliers of area_mean
desc = df_bening.area_mean.describe()
Q1 = desc[4]
Q3 = desc[6]
IQR = Q3 - Q1
lower_bound = Q1 - 1.5*IQR
upper_bound = Q3 + 1.5*IQR
print("Anything outside this area is an outlier: (", lower_bound, ",", upper_bound, ")")
df_bening[df_bening.area_mean < lower_bound].area_mean
print("Outliers: ", df_bening[(df_bening.area_mean < lower_bound) | (df_bening.area_mean > upper_bound)].area_mean.values)

Anything outside this area is an outlier: ( 118.84999999999997 , 810.45 )
Outliers:  [880.2 838.1 992.1 819.8]


* Mean ve median değerleri arasında çok büyük farklılıklar gözlenmektedir. Normal dağılıma yakınsak verilere sahip olduğumuz söylenemez. Daha fazla yorum yapmak için diğer değerleri de incelemek gerekir.
* Std değerimizin de yüksek olduğu görülmektedir.
* lower_bound olan 118.84999999999997 değerinin altında outlierımız bulunmamakta, upper_bound olan 810.45 değerimizin üzerinde 4 outlier değerimiz bulunmaktadır. 
* Outlier değerlerimizi incelediğimizde lower_bound altında hiç outlier görünmezken upper_bound üzerinde çok uç değerlere sahip olduğumuzu görüyoruz. Bu da çarpık bir dağılıma sahip olduğumuzu göstermektedir.

### **Outliers of Compactness Mean**

In [8]:
# view the values of variable compactness_mean and compactness outliers of radius_mean
desc = df_bening.compactness_mean.describe()
Q1 = desc[4]
Q3 = desc[6]
IQR = Q3 - Q1
lower_bound = Q1 - 1.5*IQR
upper_bound = Q3 + 1.5*IQR
print("Anything outside this compactness is an outlier: (", lower_bound, ",", upper_bound, ")")
df_bening[df_bening.compactness_mean < lower_bound].compactness_mean
print("Outliers: ", df_bening[(df_bening.compactness_mean < lower_bound) | (df_bening.compactness_mean > upper_bound)].compactness_mean.values)

Anything outside this compactness is an outlier: ( -0.00727499999999999 , 0.160445 )
Outliers:  [0.2233 0.1807 0.1676 0.1972 0.166  0.1836 0.2239 0.2204 0.1661]


* Mean ve median değerleri arasında çok büyük farklılıklar gözlenmemektedir. Normal dağılıma yakınsak verilere sahip olduğumuz söylenebilir.
* Std değerimizin de çok yüksek bir değere sahip olmadığını gözlemleyebilmekteyiz.
* lower_bound olan -0.00727499999999999 değerinin altında hiç outlierımız görünmemektedir, upper_bound olan 0.160445 değerimizin üzerinde 9 outlier değerimiz bulunmaktadır.
* Outlier değerlerimizi incelediğimiz zaman lower_outlier altında hiç outlier görünmezken upper_bound üzerinde çok uç değerlere sahip olduğumuzu görüyoruz. Bu da normal dağılıma yakınsak olmayan bir dataya sahip olduğumuzu gösterebilir.

## Soru 3

In [9]:
df_corr = df.loc[:,['radius_mean','area_mean','compactness_mean']]
df_corr

radius_mean  area_mean  compactness_mean
0          17.99     1001.0           0.27760
1          20.57     1326.0           0.07864
2          19.69     1203.0           0.15990
3          11.42      386.1           0.28390
4          20.29     1297.0           0.13280
..           ...        ...               ...
564        21.56     1479.0           0.11590
565        20.13     1261.0           0.10340
566        16.60      858.1           0.10230
567        20.60     1265.0           0.27700
568         7.76      181.0           0.04362

[569 rows x 3 columns]

### Correletion

In [10]:
# correletion
df_corr.corr()

radius_mean  area_mean  compactness_mean
radius_mean          1.000000   0.987357          0.506124
area_mean            0.987357   1.000000          0.498502
compactness_mean     0.506124   0.498502          1.000000

* Korelasyon matrisine göre radius_mean ile area_mean arasında yüksek pozitif korelasyon ilişkisi vardır. 
* Korelasyon matrisine göre radius_mean ile compactness_mean arasında düşük pozitif korelasyon vardır.
* Korelasyon matrisine göre area_mean ile compactness_mean arasında düşük pozitif korelasyon vardır.

### Covariance

In [11]:
# covariance
df_corr.cov()

radius_mean      area_mean  compactness_mean
radius_mean         12.418920    1224.483409          0.094197
area_mean         1224.483409  123843.554318          9.264931
compactness_mean     0.094197       9.264931          0.002789

* Kovaryans incelemelerinde 3 değerinde birbiri ile ilişkileri pozitif yönde değişim göstermektedir. 

### Pearson Correlation Coefficient

In [12]:
# pearson correlation coefficient
df_corr.corr(method = 'pearson')

radius_mean  area_mean  compactness_mean
radius_mean          1.000000   0.987357          0.506124
area_mean            0.987357   1.000000          0.498502
compactness_mean     0.506124   0.498502          1.000000

* Pearson korelasyon katsayısı metodu ile hesaplanan değerler korelasyon matrisi ile hesaplanan değerler ile benzer sonuç vermiştir.
* Pearson korelasyon matrisine göre radius_mean ile area_mean arasında yüksek pozitif korelasyon ilişkisi vardır. 
* Pearson korelasyon matrisine göre radius_mean ile compactness_mean arasında düşük pozitif korelasyon vardır.
* Pearson korelasyon matrisine göre area_mean ile compactness_mean arasında düşük pozitif korelasyon vardır.

### Spearman Rank Coefficient

In [14]:
# spearman rank coefficient
df_corr.corr(method = 'spearman')

radius_mean  area_mean  compactness_mean
radius_mean          1.000000   0.999602          0.497578
area_mean            0.999602   1.000000          0.488988
compactness_mean     0.497578   0.488988          1.000000

* Spearman rank katsayısı metodu ile hesaplanan değerler, korelasyon matrisi ile hesaplanan değerler ile ufak farklılıklar göstermiştir.
* Spearman rank katsayısına göre radius_mean ile area_mean arasında diğer yöntemlere göre daha yüksek pozitif korelasyon ilişkisi gözlemlenmiştir. 
* Spearman rank katsayısına göre radius_mean ile compactness_mean arasında diğer yöntemlere göre küçük bir oranda düşüş gözlenmiştir ve düşük pozitif korelasyon vardır.
* Spearman rank katsayısına göre area_mean ile compactness_mean arasında diğer yöntemlere göre küçük bir oranda düşüş gözlenmiştir ve düşük pozitif korelasyon vardır.

### Effect Size

In [15]:
# effect size
# eftype --> cohen / Cohen d effect size
df_rac = pg.compute_effsize(df_corr.radius_mean, df_corr.area_mean, paired = False, eftype = "cohen")
df_rcc = pg.compute_effsize(df_corr.radius_mean, df_corr.compactness_mean, paired = False, eftype = "cohen")
df_acc = pg.compute_effsize(df_corr.area_mean, df_corr.compactness_mean, paired = False, eftype = "cohen")
# eftype --> hedges / Hedges g effect size 
df_rah = pg.compute_effsize(df_corr.radius_mean, df_corr.area_mean, paired = True, eftype = "hedges")
df_rch = pg.compute_effsize(df_corr.radius_mean, df_corr.compactness_mean, paired = True, eftype = "hedges")
df_ach = pg.compute_effsize(df_corr.area_mean, df_corr.compactness_mean, paired = True, eftype = "hedges")
print("Cohen d Effect Size")
print("-Radius Mean and Area Mean with Cohen d: ",df_rac)
print("-Radius Mean and Compactness Mean: ",df_rcc)
print("-Area Mean and Compactness Mean: ",df_acc)
print("Hedges g Effect Size")
print("-Radius Mean and Area Mean with Hedges g: ",df_rah)
print("-Radius Mean and Compactness Mean: ",df_rch)
print("-Area Mean and Compactness Mean: ",df_ach)

Cohen d Effect Size
-Radius Mean and Area Mean with Cohen d:  -2.5748571535813936
-Radius Mean and Compactness Mean:  5.626829112228556
-Area Mean and Compactness Mean:  2.6313393087695855
Hedges g Effect Size
-Radius Mean and Area Mean with Hedges g:  -2.573156829685126
-Radius Mean and Compactness Mean:  5.623113398529087
-Area Mean and Compactness Mean:  2.629601686509777


* Verilen değerlerin Effect Size değerlerine ulaşmak için iki farklı yöntem kullanılmıştır.
* Cohen d ve Hedges g birbirine çok benzer yöntemlerdir. 
* Her iki yöntem ile hesaplanan effect size değerleri birbirlerine çok yakın sonuçlar vermiştir. 
* Cohen d yönteminde d değeri 0.2'den küçük ise düşük, 0.2 ve 0.8 arasında ise orta ve 0.8'den büyük ise yüksek büyüklükte etkiye sahiptir.
* Cohen d effect size yönteminde d değeri 0.8'den büyük olduğu için karşılaştırılan her değer için etki büyüklüğü kuvvetli yorumunu yapabiliriz. Bu değerlendirme sonucunda her bir verinin beraber kullanımlarının tek kullanılmalarına kıyasla daha etkili sonuçlar vereceğini söyleyebiliriz.
* Sonuçlarımızı karşılaştırdığımızda ise en büyük etkinin Radius Mean ve Compactness Mean arasında olduğu gözlenmektedir. Bu durumda bu iki verinin beraber kullanılması en etkili sonuçları verecektir.

## Soru 4

   1. Bayes teoreminin gerçek hayatta karşımıza çıkan en ünlü örneklerinden biri İkinci Dünya Savaşında İngiliz bilim insanı Alan Turing tarafından Nazilerin Enigma kodlarını kırmak için kullanılmasıdır. 
   
   * İkinci Dünya Savaşı'nda Naziler tarafından Enigma ile gönderilen mesajlarla iletişim kurulmuştur. Alan Turing, Bayes Teoremine dayanan bir sistem geliştirmiştir. Bu sistem ile gönderilen mesajlardaki bir dizi harfin tahmin edilmesi sağlanmış, olasılık hesaplamaları ile İkinci Dünya Savaşı'nın seyrinin değişmesini sağlamıştır.
   
   2. İleri yaşlardaki bireylerde ortaya çıkan otozomal dominant hastalıkların Bayes denklemi ile risk hesaplamaları yapılmaktadır. 
   
   * Örneğin Huntinton hastalığı ilk belirtileri 25-50 yaşlar arasında ortaya çıkan otozomal dominant bir hastalıktır. Diyelim ki babasını Huntinton hastalığından kaybetmiş 50 yaşında bir birey, çocuğunda bu hastalığın bulunma riskini öğrenmek istemektedir. Bu bireye babasından mutant genin kalıtılmış olma ihtimali 1/2'dir. Ancak hastalık semptonlarının 50 yaş öncesinde ortaya çıkma olasılığı %80 olduğu için örneğimizdeki bireyde hastalığın gelişme olasılığı %20 civarındadır. Bayes Teoremi kullanılarak yapılan hesaplamalar sonunda bireyin hastalanma olasılığı 1/6'dır. Çocuk babasının genlerinin yarısını aldığından mutant geni alma olasılığı 1/2 ve babanın hastalanma riski 1/6 olduğunu göz önüne aldığımızda bireyin çocuğunun hastalanma riski 1/12 olarak bulunur. Aynı zamanda babanın her hastalanmadan geçirdiği yıl, çocuğun da hastalanma olasılığını giderek azaltmaktadır. 

   3. Herhangi bir hastalığın teşhisi için geliştirilen testlerin güvenilirlik ve hassasiyet ölçümlerinde Bayes Teoremi kullanılmaktadır.
   
   * Bir testin hasta bireylerde %95 pozitif sonuç verdiğini varsayalım. İlk bakışta %95 değerine bakarak güvenilir bir test olarak nitelendirdiğimiz testlerin Bayes Teoremi ile gerekli hesaplamalar sonucunda test pozitif sonuç verdiğinde aslında testi pozitif çıkan kişinin hasta olma olasılığının %8.7 olduğunu görebiliriz. Bu sonuca baktığımız zaman testin aslında %91.3 oranında yanlış pozitif sonuç verdiğini görmüş oluruz. Böyle durumlarda ilk etapta %95 pozitif sonuç veren testin Bayes Teoremi hesaplamaları ile güvenilmez bir test olduğu tespit edilebilmektedir.
   
   4. Bayes Teoremi spam e-postaları yakalamak ve istenmeyen e-postaları filtrelemek için kullanılmaktadır. 
   
   * E- posta filtrelemelerde Bayes Teoremi kullanılarak hesaplanan Naive Bayes Spam Filtreleme yöntemi en sık kullanılan yöntemlerden biridir. Kelimeler ve bazı başka parametreleri kullanarak spam olan ve olmayan e-postaları tespit etmek, kullanıcıların  e-posta kullanım durumuna göre de  işlemler yapmaktadır.
   
   5. Doğal afet duyarlılık haritaları üretilmesinde Bayes Teoremi kullanılmaktadır.
   
   * Örneğin Artvin ili Merkez ilçesinin heyelan duyarlılık haritasınının oluşturulmasında Bayes Teoremi kullanılmıştır. 
   
   
   
   
    Not: Talbot' un olasılıkla ilgili bir sözüne denk geldim okumalarım sırasında ve çok hoşuma gitti. Sizinle de paylaşmak istedim :)
    
   >_**"Önemli konularda olasılıkları bilmek, aslında önemli konularda gerçekleri bilmekten daha çok tercih edilmelidir."**_ 
                                                                                                                   - ***Talbot*** -

## Soru 5 